compiler/rustc_mir/src/transform/instrument_coverage.rs

```diff
@@ -82,26 +93,261 @@ impl<'tcx> MirPass<'tcx> for InstrumentCoverage {
+        trace!("InstrumentCoverage starting for {:?}", mir_source.def_id());
     }
 }
 
-#[derive(Clone)]
-struct CoverageRegion {
-    pub span: Span,
+/// A BasicCoverageBlock (BCB) represents the maximal-length sequence of CFG (MIR) BasicBlocks
+/// without conditional branches.
+#[derive(Debug, Clone)]
+struct BasicCoverageBlock {
     pub blocks: Vec<BasicBlock>,
 }
 
+impl BasicCoverageBlock {
+    pub fn leader_bb(&self) -> BasicBlock {
+        self.blocks[0]
+    }
+
+    pub fn id(&self) -> String {
+        format!(
+            "@{}",
+            self.blocks
+                .iter()
+                .map(|bb| bb.index().to_string())
+                .collect::<Vec<_>>()
+                .join(ID_SEPARATOR)
+        )
+    }
+}
+
+struct BasicCoverageBlocks {
+    vec: IndexVec<BasicBlock, Option<BasicCoverageBlock>>,
+}
+
+impl BasicCoverageBlocks {
+    pub fn from_mir(mir_body: &mir::Body<'tcx>) -> Self {
+        let mut basic_coverage_blocks =
+            BasicCoverageBlocks { vec: IndexVec::from_elem_n(None, mir_body.basic_blocks().len()) };
+        basic_coverage_blocks.extract_from_mir(mir_body);
+        basic_coverage_blocks
+    }
+
+    pub fn iter(&self) -> impl Iterator<Item = &BasicCoverageBlock> {
+        self.vec.iter().filter_map(|option| option.as_ref())
+    }
+
+    fn extract_from_mir(&mut self, mir_body: &mir::Body<'tcx>) {
+        // Traverse the CFG but ignore anything following an `unwind`
+        let cfg_without_unwind = ShortCircuitPreorder::new(mir_body, |term_kind| {
+            let mut successors = term_kind.successors();
+            match &term_kind {
+                // SwitchInt successors are never unwind, and all of them should be traversed
+                TerminatorKind::SwitchInt { .. } => successors,
+                // For all other kinds, return only the first successor, if any, and ignore unwinds
+                _ => successors.next().into_iter().chain(&[]),
+            }
+        });
+
+        let mut blocks = Vec::new();
+        for (bb, data) in cfg_without_unwind
+            .filter(|(_, data)| data.terminator().kind != TerminatorKind::Unreachable)
+        {
+            if let Some(last) = blocks.last() {
+                let predecessors = &mir_body.predecessors()[bb];
+                if predecessors.len() > 1 || !predecessors.contains(last) {
+                    self.add_basic_coverage_block(blocks.split_off(0));
+                    debug!(
+                        "  because {}",
+                        if predecessors.len() > 1 {
+                            "predecessors.len() > 1".to_owned()
+                        } else {
+                            format!("bb {} is not in precessors: {:?}", bb.index(), predecessors)
+                        }
+                    );
+                }
+            }
+            blocks.push(bb);
+
+            let term = data.terminator();
+
+            // inject counters at branch blocks only
```

Comment:
It's a good question, but yes. (In fact, I tested the theory by removing the terminator check, just in case I was wrong.)

I'm adding some additional comments on this.

The predecessors rule starts a new `BasicCoverageBlock` (and closes the  non-branching `BasicBlock` sequence captured in the `blocks` vector as a separate `BasicCoverageBlock` because the current `BasicBlock` has multiple _incoming_ edges.

The terminator check closes a sequence of `BasicBlock`s with a `BasicBlock` that has more than one _outgoing_ edge *OR* a `BasicBlock` that exits the function.

I'm pretty sure both checks are required.

For example:

```rust
fn somefun() {
    println!("entered of function");
    if some_condition {
        // one incoming edge, but these statements have to be in their own BasicCoverageBlock
        println!("some_condition is true");
    }
    // multiple incoming edges here, start a new BasicCoverageBlock
}
```

The `SwitchInt` terminator has to close the current `BasicCoverageBlock`, which includes the first `println!()` and the `SwitchInt` (`if` condition check), because the next block in the traversal (in this case, the content of the `true` branch) _must_ start a new `BasicCoverage` block.

Reply:
Comments added

Reply:
Oh, is there a successors() helper like predecessors()? That'd make the symmetry a bit more obvious to me and keep less knowledge of individual terminators out of this code. As written I'm a bit worried about the wildcard here, someone could add a new TerminatorKind and forget to update this code.

Reply:
Yes, I've added comments above the CFG traversal loop to clarify this, hoping that helps. The CFG traversal uses successors() to create the preorder traversal iterator. The unwind filter (above the loop) is where successors() is actually called, so unwind paths can be filtered out.

Reply:
I also expanded the wildcard into the individual variants.